src/ints/mouse_dos_driver.cpp

```diff
@@ -0,0 +1,1440 @@
+            state.times_pressed[idx] = 0;
+        } else {
+            // unsupported - try to do something same
+            reg_ax = buttons.data;
+            reg_bx = 0;
+            reg_cx = 0;
+            reg_dx = 0;
+        }
+        break;
+    }
+    case 0x06: // MS MOUSE v1.0+ / CuteMouse - get button release data /
+               // mouse wheel data
+    {
+        const uint16_t idx = reg_bx; // button index
+        if (idx == 0xffff && state.cute_mouse) {
+            // 'magic' index for checking wheel instead of button
+            reg_bx = GetResetWheel16bit();
+            reg_cx = state.last_wheel_moved_x;
+            reg_dx = state.last_wheel_moved_y;
+        } else if (idx < NUM_BUTTONS) {
+            reg_ax = buttons.data;
+            reg_bx = state.times_released[idx];
+            reg_cx = state.last_released_x[idx];
+            reg_dx = state.last_released_y[idx];
+
+            state.times_released[idx] = 0;
+        } else {
+            // unsupported - try to do something same
+            reg_ax = buttons.data;
+            reg_bx = 0;
+            reg_cx = 0;
+            reg_dx = 0;
+        }
+        break;
+    }
+    case 0x07: // MS MOUSE v1.0+ - define horizontal cursor range
+        // Lemmings set 1-640 and wants that. iron seeds set 0-640 but
+        // doesn't like 640 Iron seed works if newvideo mode with mode
+        // 13 sets 0-639 Larry 6 actually wants newvideo mode with mode
+        // 13 to set it to 0-319
+        state.minpos_x = std::min((int16_t)reg_cx, (int16_t)reg_dx);
+        state.maxpos_x = std::max((int16_t)reg_cx, (int16_t)reg_dx);
+        // Battlechess wants this
+        pos_x = std::clamp(pos_x,
+                           static_cast<float>(state.minpos_x),
+                           static_cast<float>(state.maxpos_x));
+        // Or alternatively this:
+        // pos_x = (state.maxpos_x - state.minpos_x + 1) / 2;
+        LOG(LOG_MOUSE, LOG_NORMAL)
+        ("Define Hortizontal range min:%d max:%d", state.minpos_x, state.maxpos_x);
+        break;
+    case 0x08: // MS MOUSE v1.0+ - define vertical cursor range
+        // not sure what to take instead of the CurMode (see case 0x07
+        // as well) especially the cases where sheight= 400 and we set
+        // it with the mouse_reset to 200 disabled it at the moment.
+        // Seems to break syndicate who want 400 in mode 13
+        state.minpos_y = std::min((int16_t)reg_cx, (int16_t)reg_dx);
+        state.maxpos_y = std::max((int16_t)reg_cx, (int16_t)reg_dx);
+        // Battlechess wants this
+        pos_y = std::clamp(pos_y,
+                           static_cast<float>(state.minpos_y),
+                           static_cast<float>(state.maxpos_y));
+        // Or alternatively this:
+        // pos_y = (state.maxpos_y - state.minpos_y + 1) / 2;
+        LOG(LOG_MOUSE, LOG_NORMAL)
+        ("Define Vertical range min:%d max:%d", state.minpos_y, state.maxpos_y);
+        break;
```

Comment:
These game-specific hack are a bit worrying... Or is it that some games just rely on some undefined behaviour. which has to be kind of reverse-engineered by inspecting the behaviour of these games?

Reply:
This particular one looks like implementation not strictly following MS MOUSE behavior, discovered while testing a particular game.